service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,209 @@
+package matching
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/util"
+)
+
+// checkAssignmentConditions returns an error if the new set of assignment rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - No assignment rule's TargetBuildId can be a member of an existing version set
+// - If there existed an "unfiltered" assigment rule (which can accept any build id), at least one must still exist
+// - To override the unfiltered assignment rule requirement, the user can specify force = true
+func checkAssignmentConditions(g *persistencepb.VersioningData, maxARs int, force, hadUnfiltered bool) error {
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && countActiveAR(rules) > maxARs {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace dynamic config (%v/%v)", len(rules), maxARs))
+	}
+	if tbid, ok := isInVersionSet(rules, g.GetVersionSets()); ok {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update breaks requirement, target build id %s is already a member of version set", tbid))
+	}
+	if force == true {
+		return nil
+	}
+	if hadUnfiltered && !hasUnfiltered(rules) {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+func countActiveAR(rules []*persistencepb.AssignmentRule) int {
+	cnt := 0
+	for _, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// hasUnfiltered returns true if there exists an assignment rule with no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func hasUnfiltered(rules []*persistencepb.AssignmentRule) bool {
```

Comment:
Usually when a function checks a container of existing elements which match or do not match some rule we call function something like: anyUnfiltered, containsUnflitered etc. It could be just my preference though.

Reply:
i like `containsUnflitered`, happy to change that